beps/bep_0056.rst

```diff
@@ -0,0 +1,145 @@
+:BEP: 56
+:Title: Data compression extension
+:Version: $Revision$
+:Last-Modified: $Date$
+:Author:  Alexander Ivanov <saiv46@yandex.ru>
+:Status:  Draft
+:Type:    Standards Track
+:Created: 31-Sep-2021
+:Post-History: 
+
+Abstract
+========
+This extension adds a capability for clients to negotiate and use
+compression methods for data streams or torrent pieces, effectively
+improving bandwidth for supporting clients.
+
+Rationale
+=========
+This extension would allow clients to download files faster, without
+using file archivers. Since large files are often pre-compressed before
+torrent creation, downloaders needs to keep both the archives
+(for seeding) and uncompressed files (for own usage).
+
+Most users prefer to remove such torrents, thus harming proper file
+distribution. For example: Organizations using Bittorrent for software
+distribution needs to have centralized storage for new customers, no
+matter how many customers have the same software already.
+
+Extension header
+================
+
+This extension uses the extension protocol (specified in `BEP 0010`_)
+to advertise client capability of using chunk compression. It defines
+following items in the extension protocol handshake message:
++-------+-----------------------------------------------------------+
+| name  | description                                               |
++=======+===========================================================+
+| c     | Dictionary of supported compression algorithms which maps |
+|       | its identifiers to its priority (unsigned 8-bit integer), |
+|       | clients can adjust it based on compression speed/ratio,   |
+|       | hardware support, performance, and power mode et cetera.  |
+|       | Priority set to zero means that the compression algorithm |
+|       | is not supported or disabled by user, the client must     |
+|       | ignore unknown algorithms.                                |
++-------+-----------------------------------------------------------+
+
+
+
+The compression algorithm is selected by taking the dictionary item with
+highest priority from intersection of items supported by both peers,
+if there isn't any suitable compression algorithm - compression will be disabled.
```

Comment:
It seems like an unnecessary requirement that the same algorithm is used in both directions. It also seems like it would complicate things.

The fact that there's no message to ensure the clients agree on which algorithm is used seems risky. You don't specify how to resolve ambiguities. There may be 2 algorithms that are equally good options.

Reply:
I have taken the assumption that the same algorithm is used in both directions to simplify the negotiation process. Once both clients shared dictionaries, no further messages are required. It's unlikely that two algorithms would have the same priority on two different clients, but I should have explained it more clearly.

Reply:
> I have taken the assumption that the same algorithm is used in both directions to simplify the negotiation process.

you're making it more complicated by introducing negotiation in the first place.

> It's unlikely that two algorithms would have the same priority on two different clients

Unlikely things happen all the time, especially when you have ~100 million peers.

Reply:
> you're making it more complicated by introducing negotiation in the first place.

Yeah, but that's necessary due to clients disabling/implementing various algorithms.

> > It's unlikely that two algorithms would have the same priority on two different clients
>
> Unlikely things happen all the time, especially when you have ~100 million peers.

Tried to resolve this by taking TLS approach, now in `crequest` client will enumerate what algorithms it's capable of, and then having other client to respond in `cresponse` with selected algorithms to send and receive.